kb/development/developer-builds.md

```diff
@@ -0,0 +1,727 @@
+```
+
+or
+
+```
+mbedtls-prepare-build -d build-clang-c11 --cc clang --cflags '-O2 -std=c11'
+```
+
+#### Set up a debug build
+
+With CMake:
+
+```
+cmake -DCMAKE_BUILD_TYPE=Debug -B build-debug
+```
+
+With mbedtls-prepare-build (the build directory defaults to `build-debug`):
+
+```
+mbedtls-prepare-build -p debug
+```
+
+#### Set up a build with ASan and UBSan
+
+With CMake:
+
+```
+cmake -DCMAKE_BUILD_TYPE=Asan .. -B build-debug
+```
+
+With mbedtls-prepare-build (the build directory defaults to `build-debug`):
+
+```
+mbedtls-prepare-build -p asan
+```
+
+In both cases, the command line only mentions ASan (address sanitizer), but the build scripts also enable UBSan (undefined behavior sanitizer).
+
+#### Set up a build with code size optimization
+
+Both CMake and mbedtls-prepare-build default to `-Os` in `CFLAGS` if no build type is specified. See “[Choose compiler and compiler flags manually](#choose-compiler-and-compiler-flags-manually)” if you want a different choice.
+
+With mbedtls-prepare-build, you can reproduce `component_build_arm_none_eabi_gcc_m0plus` from `all.sh` as follows:
+
+```
+mbedtls-prepare-build -p m0plus
+```
+
+You may be interested in [`mbedtls-size-dwim`](https://github.com/Mbed-TLS/mbedtls-docs/blob/main/tools/bin/mbedtls-size-dwim) (located alongside `mbedtls-prepare-build`) to track code size changes across successive commits.
+
+#### Set up a build with code size optimization
+
+With CMake, the build type `Release` sets the optimization level to `-O2` (`-Ohz` with IAR):
+
+```
+cmake -DCMAKE_BUILD_TYPE=Release -B build-opt
+```
+
+With mbedtls-prepare-build, the preset `opt` sets the optimization level to `-O3` (the build directory defaults to `build-opt`):
+
+```
+mbedtls-prepare-build -p opt
+```
+
+See “[Choose compiler and compiler flags manually](#choose-compiler-and-compiler-flags-manually)” if you want a different choice.
+
+#### Set up a build for profiling
+
+This section explains how to set up a build directory for profiling with GProf.
+A profiling build requires specific compiler and linker options, so it is generally easier to set it up in a separate directory.
+
+With CMake:
+
+```
+CFLAGS='-O2 -pg' LDFLAGS='-pg' cmake -B build-prof
+```
+
+With mbedtls-prepare-build, doing normal runs for `.run` targets by default:
+
+```
+mbedtls-prepare-build -d build-prof --cflags='-O2 -pg' --ldflags='-pg'
+```
+
+Then see “[Run tests with profiling](run-tests-with-profiling)”.
+
+#### Set up a build for test coverage
+
+This section explains how to set up a build directory for measuring test coverage with lcov.
+Measuring test coverage requires specific compiler and linker options, so it is generally easier to set it up in a separate directory.
+
+With CMake:
+
+```
+cmake -B build-coverage CMAKE_BUILD_TYPE=Coverage
+```
+
+With mbedtls-prepare-build:
+
+```
+mbedtls-prepare-build -p coverage
+```
+
+### Library configuration setup
+
+#### Build in the full configuration
+
+With CMake: ???
```

Comment:
Pending in https://github.com/Mbed-TLS/mbedtls/pull/10038.

This is something I do very often. It's currently broken with mbedtls-prepare-build in Mbed TLS, because mbedtls-prepare-build needs some refactoring to handle more than one config file. I expect I'll fix that soonish.